### PR TITLE
Add "axis" and "axis_w" arguments in FC to support customized axix to reduce dim.

### PR DESCRIPTION
**Pull request by gujinghui**
Add "axis" and "axis_w" arguments in FC to support customized axix to reduce dim.

### CODE REVIEW

**Review by yinghai — COMMENTED** [review]
The title of the PR is misleading. For general FC op, we already support it. 

This is about Nomnigraph transformation and support in IDEEP, right?

**Comment by gujinghui** [comment]
@yinghai 
No. Previously, we only support fixed axis to reduce the dims in FC, for example, 4D (nchw) to 2D (nC).
Now, we can support mutable axis in FC.

Yes, the title is misleading. Will change it.

**Comment by gujinghui** [comment]
hi @yinghai ,

Regardless of #12970 and #9488, can we merge this PR?
This one is not only for pre-converting filters.
To support customized axis in FC, we need this PR.

File: caffe2/ideep/operators/fully_connected_op.cc [hunk 1]
```diff
@@ -2,67 +2,111 @@
...
+        axis_w_(OperatorBase::GetSingleArgument<int32_t>("axis_w", 1)),
         float16_compute_(
             OperatorBase::GetSingleArgument<bool>("float16_compute", false)) {}
   virtual ~IDEEPFullyConnectedOp() {}
 
   bool RunOnDevice() override {
-    const auto& X = Input(INPUT);
-    const auto& filter = Input(FILTER);
-    auto* Y = Output(OUTPUT);
+    const auto &X = Input(INPUT);
```
Comment (yinghai) [inline caffe2/ideep/operators/fully_connected_op.cc]: Can we not change the format back to `auto &X`?  Our coding style is `auto& X`.

**Review by yinghai — CHANGES_REQUESTED** [review]

File: caffe2/ideep/operators/fully_connected_op.cc (hunk 1 above)
Comment (gujinghui) [reply caffe2/ideep/operators/fully_connected_op.cc]: will fix it soon

**Comment by gujinghui** [comment]
@yinghai 
pls help review.

**Review by yinghai — CHANGES_REQUESTED** [review]
I think this needs rebase.

**Comment by gujinghui** [comment]
@yinghai 
rebased. pls have a try.

**Comment by yinghai** [comment]
Could you fix the clang format too? I found that your style seems to be consistently different from ours. :0

**Comment by gujinghui** [comment]
> Could you fix the clang format too? I found that your style seems to be consistently different from ours. :0

Sure. Done. Pls help review.
BTW, the format file is <root_path>/caffe2/.clang-format, other than <root_path>/.clang-format, right?

**Comment by gujinghui** [comment]
rebased to latest.
@yinghai 
please help merge it.

**Comment by yinghai** [comment]
@gujinghui Do we have any tests for this change?

**Comment by gujinghui** [comment]
@yinghai 
yes. We have test cases for this in caffe2/python/ideep/fc_op_test.py

rebased to latest.

File: caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h [hunk 2]
```diff
@@ -565,13 +565,31 @@ class Clip : public NeuralNetOperator {
...
 class FC : public NeuralNetOperator {
  public:
-  FC() : NeuralNetOperator(NNKind::FC) {}
+  FC() : NeuralNetOperator(NNKind::FC), Axis(1), Axis_w(1) {}
 
   ~FC() {}
 
   NOMNIGRAPH_DEFINE_NN_RTTI(FC);
 
+  int getAxis() const {
```
Comment (yinghai) [inline caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h]: This won't work as this is generated file. You have to change: 
https://github.com/pytorch/pytorch/blob/265c97decfa531810f3f5a54b3c96e6ffc2fd284/caffe2/core/nomnigraph/ops.def#L1-L7

**Review by yinghai — CHANGES_REQUESTED** [review]

File: caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h (hunk 2 above)
Comment (gujinghui) [reply caffe2/core/nomnigraph/include/nomnigraph/Generated/OpClasses.h]: @yinghai 
You're right. My mistake.
Fixed. Pls help review.

**Comment by gujinghui** [comment]
@yinghai 
Fixed. Pls help review.

**Pull request closed without merging**